File: behavioral/chain-of-responsibility.cpp
#include <iostream>
#include <memory>
#include <string>

// Abstract Handler
class SupportHandler
{
public:
  virtual ~SupportHandler() = default;

  void setNext(std::shared_ptr<SupportHandler> handler)
  {
    nextHandler = handler;
  }

  virtual void handleRequest(const std::string &severity)
  {
    if (nextHandler)
    {
      nextHandler->handleRequest(severity);
    }
  }

protected:
  std::shared_ptr<SupportHandler> nextHandler;
};

// Concrete Handler: LowLevelSupport
class LowLevelSupport : public SupportHandler
{
public:
  void handleRequest(const std::string &severity) override
  {
    if (severity == "low")
    {
      std::cout << "LowLevelSupport: Handling low severity ticket" << std::endl;
    }
    else if (nextHandler)
    {
      nextHandler->handleRequest(severity);
    }
  }
};

// Concrete Handler: MidLevelSupport
class MidLevelSupport : public SupportHandler
{
public:
  void handleRequest(const std::string &severity) override
  {
    if (severity == "medium")
    {
      std::cout << "MidLevelSupport: Handling medium severity ticket" << std::endl;
    }
    else if (nextHandler)
    {
      nextHandler->handleRequest(severity);
    }
  }
};

// Concrete Handler: HighLevelSupport
class HighLevelSupport : public SupportHandler
{
public:
  void handleRequest(const std::string &severity) override
  {
    if (severity == "high")
    {
      std::cout << "HighLevelSupport: Handling high severity ticket" << std::endl;
    }
    else if (nextHandler)
    {
      nextHandler->handleRequest(severity);
    }
  }
};

// Example usage
int main()
{
  auto lowLevel = std::make_shared<LowLevelSupport>();
  auto midLevel = std::make_shared<MidLevelSupport>();
  auto highLevel = std::make_shared<HighLevelSupport>();

  lowLevel->setNext(midLevel);
  midLevel->setNext(highLevel);

  lowLevel->handleRequest("low");    // Output: LowLevelSupport: Handling low severity ticket
  lowLevel->handleRequest("medium"); // Output: MidLevelSupport: Handling medium severity ticket
  lowLevel->handleRequest("high");   // Output: HighLevelSupport: Handling high severity ticket

  return 0;
}

File: behavioral/chain-of-responsibility.ts
// Abstract Handler
abstract class SupportHandler {
  protected nextHandler: SupportHandler | null = null;

  public setNext(handler: SupportHandler): SupportHandler {
    this.nextHandler = handler;
    return handler;
  }

  public handleRequest(severity: string): void {
    if (this.nextHandler) {
      this.nextHandler.handleRequest(severity);
    }
  }
}

// Concrete Handler: LowLevelSupport
class LowLevelSupport extends SupportHandler {
  public handleRequest(severity: string): void {
    if (severity === "low") {
      console.log("LowLevelSupport: Handling low severity ticket");
    } else if (this.nextHandler) {
      this.nextHandler.handleRequest(severity);
    }
  }
}

// Concrete Handler: MidLevelSupport
class MidLevelSupport extends SupportHandler {
  public handleRequest(severity: string): void {
    if (severity === "medium") {
      console.log("MidLevelSupport: Handling medium severity ticket");
    } else if (this.nextHandler) {
      this.nextHandler.handleRequest(severity);
    }
  }
}

// Concrete Handler: HighLevelSupport
class HighLevelSupport extends SupportHandler {
  public handleRequest(severity: string): void {
    if (severity === "high") {
      console.log("HighLevelSupport: Handling high severity ticket");
    } else if (this.nextHandler) {
      this.nextHandler.handleRequest(severity);
    }
  }
}

// Example usage
const lowLevel = new LowLevelSupport();
const midLevel = new MidLevelSupport();
const highLevel = new HighLevelSupport();

lowLevel.setNext(midLevel).setNext(highLevel);

lowLevel.handleRequest("low");
lowLevel.handleRequest("medium");
lowLevel.handleRequest("high");
